constraint/solver/gkrgates/registry.go

```diff
@@ -0,0 +1,272 @@
+					return fmt.Errorf("for gate %s: %v", s.name, err)
+				}
+			}
+		}
+
+		if s.solvableVar == -1 {
+			if !s.noSolvableVarVerification { // find a solvable variable
+				s.solvableVar = gateVer.findSolvableVar(f, nbIn)
+			}
+		} else {
+			// solvable variable given
+			if !s.noSolvableVarVerification && !gateVer.isVarSolvable(f, s.solvableVar, nbIn) {
+				return fmt.Errorf("cannot verify the solvability of variable %d in gate %s", s.solvableVar, s.name)
+			}
+		}
+
+	}
+
+	gatesLock.Lock()
+	defer gatesLock.Unlock()
+	gates[s.name] = gkrtypes.NewGate(f, nbIn, s.degree, s.solvableVar)
+	return nil
+}
+
+func Get(name gkr.GateName) *gkrtypes.Gate {
+	gatesLock.Lock()
+	defer gatesLock.Unlock()
+	if gate, ok := gates[name]; ok {
+		return gate
+	}
+	panic(fmt.Sprintf("gate \"%s\" not found", name))
+}
+
+type gateVerifier struct {
+	isAdditive   func(f gkr.GateFunction, i int, nbIn int) bool
+	findDegree   func(f gkr.GateFunction, max, nbIn int) (int, error)
+	verifyDegree func(f gkr.GateFunction, claimedDegree, nbIn int) error
+}
+
+func NewGateVerifier(curve ecc.ID) (*gateVerifier, error) {
+	var (
+		o   gateVerifier
+		err error
+	)
+	switch curve {
+	case ecc.BLS12_377:
+		o.isAdditive = bls12377.IsGateFunctionAdditive
+		o.findDegree = bls12377.FindGateFunctionDegree
+		o.verifyDegree = bls12377.VerifyGateFunctionDegree
+	case ecc.BLS12_381:
+		o.isAdditive = bls12381.IsGateFunctionAdditive
+		o.findDegree = bls12381.FindGateFunctionDegree
+		o.verifyDegree = bls12381.VerifyGateFunctionDegree
+	case ecc.BLS24_315:
+		o.isAdditive = bls24315.IsGateFunctionAdditive
+		o.findDegree = bls24315.FindGateFunctionDegree
+		o.verifyDegree = bls24315.VerifyGateFunctionDegree
+	case ecc.BLS24_317:
+		o.isAdditive = bls24317.IsGateFunctionAdditive
+		o.findDegree = bls24317.FindGateFunctionDegree
+		o.verifyDegree = bls24317.VerifyGateFunctionDegree
+	case ecc.BN254:
+		o.isAdditive = bn254.IsGateFunctionAdditive
+		o.findDegree = bn254.FindGateFunctionDegree
+		o.verifyDegree = bn254.VerifyGateFunctionDegree
+	case ecc.BW6_633:
+		o.isAdditive = bw6633.IsGateFunctionAdditive
+		o.findDegree = bw6633.FindGateFunctionDegree
+		o.verifyDegree = bw6633.VerifyGateFunctionDegree
+	case ecc.BW6_761:
+		o.isAdditive = bw6761.IsGateFunctionAdditive
+		o.findDegree = bw6761.FindGateFunctionDegree
+		o.verifyDegree = bw6761.VerifyGateFunctionDegree
+	default:
+		err = fmt.Errorf("unsupported curve %s", curve)
+	}
+	return &o, err
+}
+
+// GetDefaultGateName provides a standardized name for a gate function, depending on its package and name.
+// NB: For anonymous functions, the name is the same no matter the implicit arguments provided.
+func GetDefaultGateName(fn gkr.GateFunction) gkr.GateName {
+	// copied from solver.GetHintName
+	fnptr := reflect.ValueOf(fn).Pointer()
+	name := runtime.FuncForPC(fnptr).Name()
+	return gkr.GateName(newToOldStyle(name))
+}
+
+func newToOldStyle(name string) string {
```

Comment:
We don't need it here I think. It was only for compatibility when we upgraded from Go 1.18 imo. But because the APIs have changed anyway, then we don't need to consider to be backwards compatible now as now we force Go 1.23 anyway.